src/middlewared/middlewared/plugins/asigra.py

```diff
@@ -0,0 +1,358 @@
+            return exists
+
+        exists = False
+        try:
+            exists = (cur.fetchone()[0] != 0)
+        except Exception:
+            exists = False
+
+        cur.close()
+        con.close()
+
+        return exists
+
+    async def asigra_database_create(self):
+        con = None
+
+        try:
+            con = pg.connect("dbname='template1' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to template1: {}".format(e))
+            return False
+
+        con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        cur = con.cursor()
+
+        try:
+            self.logger.debug("there is no dssystem database found in the postgres database. Creating ...")
+            cur.execute("CREATE DATABASE dssystem")
+
+        except Exception as e:
+            self.logger.error("Can't create dssystem database: {}".format(e))
+            return False
+
+        cur.close()
+        con.close()
+
+        return True
+
+    async def asigra_database_init(self):
+        con = None
+
+        try:
+            con = pg.connect("dbname='dssystem' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to dssystem: {}".format(e))
+            return False
+
+        con.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+        cur = con.cursor()
+
+        with open(os.path.join(self.dssystem_db_path, "postgresdssystem.sql"), "r") as f:
+            try:
+                cur.execute(f.read())
+
+            except Exception as e:
+                self.logger.error("Can't init dssystem database: {}".format(e))
+                return False
+
+        with open(os.path.join(self.dssystem_db_path, "dssystem_locale_postgres.sql"), "r") as f:
+            try:
+                cur.execute(f.read())
+
+            except Exception as e:
+                self.logger.error("Can't init dssystem database: {}".format(e))
+                return False
+
+        cur.close()
+        con.close()
+
+        return True
+
+    async def get_db_number(self):
+        db_number = 0
+        con = None
+
+        try:
+            con = pg.connect("dbname='dssystem' user='{}'".format(self.pg_user))
+
+        except Exception as e:
+            self.logger.error("Can't connect to dssystem: {}".format(e))
+            return db_number
+
+        cur = con.cursor()
+        try:
+            cur.execute("SELECT db_number FROM ds_data")
+
+        except Exception as e:
+            self.logger.error("Can't get db_number: {}".format(e))
+            return db_number
+
+        try:
+            db_number = cur.fetchone()[0]
+
+        except Exception as e:
+            self.logger.error("Can't get db_number results: {}".format(e))
+            return db_number
+
+        cur.close()
+        con.close()
+
+        return db_number
+
+    async def asigra_database_update(self):
```

Comment:
Given all the blocking operations (database) this should not be a coroutine, but a normal python method.